lib/block-supports/typography.php

```diff
@@ -92,117 +92,142 @@ function gutenberg_apply_typography_support( $block_type, $block_attributes ) {
 		$has_line_height = isset( $block_attributes['style']['typography']['lineHeight'] );
 		if ( $has_line_height ) {
-			$styles[] = sprintf( 'line-height: %s;', $block_attributes['style']['typography']['lineHeight'] );
+			$styles['lineHeight'] = $block_attributes['style']['typography']['lineHeight'];
 		}
 	}
 
-	if ( $has_text_decoration_support && ! gutenberg_should_skip_block_supports_serialization( $block_type, 'typography', 'textDecoration' ) ) {
-		$text_decoration_style = gutenberg_typography_get_css_variable_inline_style( $block_attributes, 'textDecoration', 'text-decoration' );
-		if ( $text_decoration_style ) {
-			$styles[] = $text_decoration_style;
+	if ( $has_text_decoration_support && ! $should_skip_text_decoration ) {
+		$text_decoration       = _wp_array_get( $block_attributes, array( 'style', 'typography', 'textDecoration' ), null );
+		$text_decoration_value = gutenberg_typography_get_preset_inline_style_value( $text_decoration, 'text-decoration' );
+		if ( $text_decoration_value ) {
+			$styles['textDecoration'] = $text_decoration_value;
 		}
 	}
 
-	if ( $has_text_transform_support && ! gutenberg_should_skip_block_supports_serialization( $block_type, 'typography', 'textTransform' ) ) {
-		$text_transform_style = gutenberg_typography_get_css_variable_inline_style( $block_attributes, 'textTransform', 'text-transform' );
-		if ( $text_transform_style ) {
-			$styles[] = $text_transform_style;
+	if ( $has_text_transform_support && ! $should_skip_text_transform ) {
+		$text_transform       = _wp_array_get( $block_attributes, array( 'style', 'typography', 'textTransform' ), null );
+		$text_transform_value = gutenberg_typography_get_preset_inline_style_value( $text_transform, 'text-transform' );
+		if ( $text_transform_value ) {
+			$styles['textTransform'] = $text_transform_value;
 		}
 	}
 
-	if ( $has_letter_spacing_support && ! gutenberg_should_skip_block_supports_serialization( $block_type, 'typography', 'letterSpacing' ) ) {
-		$letter_spacing_style = gutenberg_typography_get_css_variable_inline_style( $block_attributes, 'letterSpacing', 'letter-spacing' );
-		if ( $letter_spacing_style ) {
-			$styles[] = $letter_spacing_style;
+	if ( $has_letter_spacing_support && ! $should_skip_letter_spacing ) {
+		$letter_spacing       = _wp_array_get( $block_attributes, array( 'style', 'typography', 'letterSpacing' ), null );
+		$letter_spacing_value = gutenberg_typography_get_preset_inline_style_value( $letter_spacing, 'letter-spacing' );
+		if ( $letter_spacing_value ) {
+			$styles['letterSpacing'] = $letter_spacing_value;
 		}
 	}
 
-	if ( ! empty( $classes ) ) {
-		$attributes['class'] = implode( ' ', $classes );
+	$style_engine  = gutenberg_get_style_engine();
+	$inline_styles = $style_engine->generate(
+		array( 'typography' => $styles ),
+		array(
+			'inline' => true,
+		)
+	);
+
+	$classnames = $style_engine->get_classnames(
+		array( 'typography' => $classes ),
+		array(
+			'use_schema' => true,
+		)
+	);
```

Comment:
> Seeing these two calls side-by-side, I was wondering if we should consolidate these calls so that get_classnames becomes an implementation detail of generate within the style engine? 

I agree it's something we probably should do. 

I spent some time thinking about this before I landed on two methods, and I got stuck on the structure of the arguments signature more than the return. That is, how show we should tell the style engine to deal with the different values?

The `$block_styles` arg matches Gutenberg's style object and has a very particular structure, and I couldn't see how classname slugs would fit into that. 

We could always expand the options array. I did try that but wasn't satisfied with my approach as it appeared haphazard and unintuitive, and not related at all to the main argument `$block_styles`.

```php
generate( 
     $block_styles,
      [ classnames => $array_of_classnames ]
)
```

The first argument could be a collection of `[ $block_styles, $block_classnames ]` I suppose. 

Another thought I had was extended the style value with metadata, which would mean that style engine had its own implementation of `$block_styles`. For example:

```
// Classname/slug
$style['color'] = [ 'classname' => 'crimson'  ]; 

// Regular style
$style['color'] = '#fff'; // or  [ 'value' => '#fff'  ];
``` 

The reason why I left it at two was so we could have separate methods we could ultimately combine into a single `generate`.

So maybe the current `generate` should rather be named `get_styles` or something and we create a new `generate` that calls `get_styles` or `get_classnames` accordingly.  🤷 

What do you reckon?

Reply:
Oh, that _is_ an interesting challenge! Thanks for explaining the logic there, and apologies when I was reading this yesterday, I totally missed that the following lines are grabbing root level attributes, not values within the style object 😅:

```
$classes['fontSize'] = _wp_to_kebab_case( $block_attributes['fontSize'] );
$classes['fontFamily'] = _wp_to_kebab_case( $block_attributes['fontFamily'] );
```

I think we're now bumping into one of the tricky areas where there's a part of the feature here that we probably need to decide between "do we change how it works now, or later?".

As in the border PR (https://github.com/WordPress/gutenberg/pull/37770) I think there's been discussion in various places about whether these root level attributes will (long-term) better belong within the `style` object. Ideally, we'd have everything we need right there in the style object, whether that style object is coming from block attributes, or global styles.

However, we're not there yet!

So in the spirit of the approach we've been taking thus far, which is — let's see if we can refactor what's currently working (pretty much as-is) and then iteratively switch over to new approaches — I reckon we go in the direction you've been exploring 👍

> We could always expand the options array. I did try that but wasn't satisfied with my approach as it appeared haphazard and unintuitive, and not related at all to the main argument $block_styles.

You're right, it's tricky to come up with something that feels intuitive. I was kinda wondering if we could nearly have a value in the options array called `attributes` or `additionalAttributes` or something, which is a way to pass in attributes that are not part of the `style` object — but that could feel a bit awkward, too...

Basically, I was thinking of something like:

```
$style_engine->generate(
		array( 'typography' => $styles ),
		array(
			'inline' => true,
			'attributes' => array(
				'fontSize' => 'large', // just the raw attribute, let the style engine do the kebab-ing?
				'fontFamily' => 'papyrus'
			)
		)
	);
```

But I agree, it does feel awkward including those attributes separately in the list of options, so it might not be the right approach...

> Another thought I had was to extend the style value with metadata, which would mean the style engine had its own implementation of $block_styles

That's actually not such a bad idea 🤔😀  — if we think of this as an opportunity to come up with the structure of how we long-term hope to structure the style object so that it contains those root level attributes, if we go with this approach, it might be a good way to confidence check what that shape will eventually look like.

It could also mean we kinda have an implementation which is "if the values are all in the style object, we don't need to look for anything else, otherwise, migrate over those other attributes to the style object for backwards compat" — that way we might be able to handle the desired location in the style object without ever having to worry about migrating old attributes? Just a thought....

> The reason why I left it at two methods was so we could indeed have separate methods that we could ultimately call in a single generate.
> So maybe the current generate should rather be named get_styles or whatever, and we create a new generate that calls get_styles or get_classnames accordingly.

I really like this idea, it means we've got two inner methods that are doing different things, and we have `generate` as an abstraction that sits just above, so we can have a single entry point to generating styles 👍